crates/ruff_server/src/server/api/requests/execute_command.rs

```diff
@@ -0,0 +1,144 @@
+use std::collections::HashMap;
+
+use crate::server;
+use crate::server::api::LSPResult;
+use crate::server::client;
+use crate::server::schedule::Task;
+use crate::session::Session;
+use crate::DIAGNOSTIC_NAME;
+use lsp_server::ErrorCode;
+use lsp_types::{self as types, request as req};
+use serde::Deserialize;
+
+enum Command {
+    Format,
+    FixAll,
+    OrganizeImports,
+}
+
+pub(crate) struct ExecuteCommand;
+
+#[derive(Deserialize)]
+struct TextDocumentArgument {
+    uri: types::Url,
+}
+
+impl super::RequestHandler for ExecuteCommand {
+    type RequestType = req::ExecuteCommand;
+}
+
+impl super::SyncRequestHandler for ExecuteCommand {
+    fn run(
+        session: &mut Session,
+        _notifier: client::Notifier,
+        requester: &mut client::Requester,
+        params: types::ExecuteCommandParams,
+    ) -> server::Result<Option<serde_json::Value>> {
+        let Some(command) = Command::from_str(&params.command) else {
+            return Err(anyhow::anyhow!("")).with_failure_code(ErrorCode::InvalidParams);
+        };
```

Comment:
nit: we could possibly implement `FromStr` for `Command` and then use `?` operator to avoid creating an empty `anyhow` error. I'm not sure how will `with_failure_code` comes from which I believe you're aware of.

```rs
impl FromStr for Command {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(match command {
            "ruff.applyAutofix" => Self::FixAll,
            "ruff.applyFormat" => Self::Format,
            "ruff.applyOrganizeImports" => Self::OrganizeImports,
            _ => return Err(()),
        })
    }
}
```

Reply:
Good catch. I think we should always avoid anyhow error.

Reply:
We'd still need to define an error type for `FromStr` - I think using anyhow as an error makes sense here because it lets us pass an arbitrary error message. The message being empty here was just an oversight.